design-proposals/klip-12-pull-high-availability.md

```diff
@@ -0,0 +1,323 @@
+# KLIP 12 - Implement High-Availability for Pull queries
+
+**Author**: Vicky Papavasileiou, Vinoth Chandar | 
+**Release Target**: 5.5 | 
+**Status**: _In Discussion_
+**Discussion**: _link to the design discussion PR_
+
+**tl;dr:** Enables high-availability for Ksql pull queries, in case of server failures. Current 
+design for handling failure of an active ksql server, incurs an unavailability period of several 
+seconds (10+) to few minutes due to the Streams rebalancing procedure. This work is based on new 
+Kafka Streams APIs, introduced as a part of [KIP-535](https://cwiki.apache.org/confluence/display/KAFKA/KIP-535%3A+Allow+state+stores+to+serve+stale+reads+during+rebalance), 
+which allow serving stale data from standby replicas to achieve high availability, while providing 
+eventual consistency. 
+
+## Motivation and background
+Stateful persistent queries persist their state in state stores. These state stores are partitioned
+and replicated across a number of standbys for faster recovery in case of 
+[failures](https://docs.confluent.io/current/streams/architecture.html#fault-tolerance).
+A KSQL server may host partitions from multiple state stores serving as the active host for some partitions
+and the standby for others. Without loss of generality, we will focus on one partition and one state 
+store for the remainder of this document. The active KSQL server is the server that hosts the 
+active partition whereas the standby servers are the ones that host the replicas.  
+
+Assume we have a load balancer (LB) and a cluster of three KSQL servers (`A`, `B`, `C`) where `A` 
+is the active server and `B`, `C` are the standby replicas. `A` receives updates from the source topic 
+partition and writes these updates to its state store and changelog topic. The changelog topic is 
+replicated into the state stores of servers `B` and `C`. 
+
+Now assume  LB receives a pull query (1) and sends the query to `B` (2). `B` determines that `A` 
+is the active server and forwards the request to `A` (3). `A` executes the query successfully and 
+returns the response.
```

Comment:
Is the response send directly to the client by `A` or does the reply flow back via `B` ?

Reply:
Flows back via `B` .. B acts as a the proxy for that query.